### ISSUE TITLE
more prior art

### ISSUE THREAD

**Issue by michaelficarra** [tc39-transfer/proposal-random-functions#8]
JSVerify is another property-based (generative) testing library for JavaScript. Its documentation lists its Arbitrary instances starting here: https://github.com/jsverify/jsverify#primitive-arbitraries

I think it would be a good (necessary but not sufficient) evaluation criteria to ensure that the Arbitrary instances provided by these generative testing libraries could be relatively easily defined in terms of the functions provided in this proposal.

**Comment by tabatkins** [comment]
Yup, between "random number in a range", "random int in a range", and "random value from a list", it looks like all the `jsverify` primitives are covered.

**Comment by michaelficarra** [comment]
I dunno about that. For example, it's non-trivial to get a good distribution of BigInts. You'd have to write something like this today:

```js
BigInt(function(){
  let s = '';
  do { s += Math.floor(Math.random()*10); } while(Math.random() < 0.9);
  return s;
}())
```

There's plenty of ways we could make this better:

* a built-in random BigInt generator with non-linear distributions (obviously)
* random Boolean with given probability
* random Numbers/integers with non-linear distributions
* random String given a code point generator

And there's plenty of other tricky things. There's lots of design choices to be made with random strings. It's hard to pick a Date (or Temporal.PlainDateTime in modern code) that's bounded by a year/month/day. While the web platform has `crypto.randomUUID()`, it doesn't have one that would work with a seeded PRNG.

**Comment by bakkot** [comment]
I don't think those are problems this proposal should aim to solve, personally.

**Comment by tabatkins** [comment]
There is no uniform distribution over the integers, but as long as you have a range that you *do* care about, you can get one easily. For anything else, what distribution you want will depend on your use-case pretty severely.

**Comment by michaelficarra** [comment]
Yes, use cases will vary quite a bit, but I think just a few very common (parameterisable) distributions will cover the large majority of non-uniform use cases. As you know, there's precedent for this in Java, Python, and probably others if I went looking.

This is getting a bit off-topic from the original post. Maybe I should move it to #6?

**Comment by kgryte** [comment]
Currently, the README suggests adding various distributions. I'm biased, but, in terms of prior art, I would suggest seeing how we do things in stdlib (e.g., https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/random/base), where we already have a growing collection of statistical distributions, with support for seeding and shared state.

Personally, I am not of the opinion that these sorts of functions should be standardized, especially as some distributions would require vendors to implement notoriously involved special functions (e.g., [`betainc`](https://github.com/stdlib-js/stdlib/tree/f63331919c71fc22a6022ef6df2151d0d2ef3aab/lib/node_modules/%40stdlib/math/base/special/kernel-betainc)), which typically require specialized libraries, but opinions may differ.
